### ISSUE TITLE
Project.sync should fail when stale output cannot be deleted

### ISSUE THREAD

**Issue by Jonatha1983** [gradle/gradle#8727]
When project.sync is execute - and destination directory contains files to be remove the operation succeed even though extra  files are not removed. 

### Expected Behavior
When sync cannot remove extra file in destination it should fail. 

### Current Behavior
Sync complete and workerResult api give work is done - when work was half done ( only copy without remove) 
### Context
All the sync tasks are not working as expected - lot of runtime exceptions

### Steps to Reproduce (for bugs)
Create a project with the following task: 
```

tasks.register("aFailedSync") {

    inputs.file project.buildDir.toPath().resolve("myProject.jar")

    outputs.dir project.buildDir.toPath().resolve("newFolder")

    doLast {

        def wk = project.sync {
            from project.buildDir.toPath().resolve("myProject.jar")
            into project.buildDir.toPath().resolve("newFolder")
        }
        println "wk.didWork ${wk.didWork}"
    }
}
```

create a file  myProject.jar under build folder and one under newFolder (the one that should be remove - keep it lock during execution) 
### Your Environment
https://gradle.com/s/m5ofmnjuuvlsc

**Comment by Jonatha1983** [comment]
In addition:
The page "Authoring Tasks" says: 
"Note that if a task has an output directory specified, any files added to that directory since the last time it was executed are ignored and will NOT cause the task to be out of date. This is so unrelated tasks may share an output directory without interfering with each other. If this is not the behavior you want for some reason, consider using TaskOutputs.upToDateWhen(groovy.lang.Closure) " 

Since it is related to the Sync Task I wanted to two questions: 
a) Is it possible to add to the Sync Task a Boolean property to make it run when the destination contains extra files. 
b) When a user set both the **inputs + outputs** and define an action for the **upToDateWhen** is it overriding the inputs/outputs incremental behavior or expanding it?  I looked for some documentation about it both in Docs and in API and could not found any.

```
inputs.files aFile
outputs.dir aDir
outputs.upToDateWhen { some logic to make it run on extra files in directory}
```
Thanks

**Comment by big-guy** [comment]
I found that we were inadvertently hiding the failure when deleting a single file vs a directory.  This will be fixed in 5.4.

> a) Is it possible to add to the Sync Task a Boolean property to make it run when the destination contains extra files.

I don't think we'd do this as a boolean/opt-in sort of thing, but just as something we'd always do.  How would extra files get into the directory?

> b) When a user set both the inputs + outputs and define an action for the upToDateWhen is it overriding the inputs/outputs incremental behavior or expanding it? I looked for some documentation about it both in Docs and in API and could not found any.

It adds to it.  Returning `true` from this doesn't override Gradle's up-to-date checks.
